sympy/functions/special/bessel.py

```diff
@@ -690,51 +690,70 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         return super(besselk, self)._eval_as_leading_term(x, logx, cdir)
 
     def _eval_nseries(self, x, n, logx, cdir=0):
-        # Refer https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/
-        # for more information on nseries expansion of besselk function.
         from sympy.series.order import Order
         nu, z = self.args
 
-        # In case of powers less than 1, number of terms need to be computed
-        # separately to avoid repeated callings of _eval_nseries with wrong n
         try:
             _, exp = z.leadterm(x)
         except (ValueError, NotImplementedError):
             return self
 
-        if exp.is_positive and nu.is_integer:
-            newn = ceiling(n/exp)
-            bn = besseli(nu, z)
-            a = ((-1)**(nu - 1)*log(z/2)*bn)._eval_nseries(x, n, logx, cdir)
-
-            b, c = [], []
-            o = Order(x**n, x)
-            r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
-            if r is S.Zero:
-                return o
-            t = (_mexpand(r**2) + o).removeO()
-
-            if nu > S.Zero:
-                term = r**(-nu)*factorial(nu - 1)/2
-                b.append(term)
-                for k in range(1, nu):
-                    denom = (k - nu)*k
-                    if denom == S.Zero:
-                        term *= t/k
-                    else:
-                        term *= t/denom
-                    term = (_mexpand(term) + o).removeO()
+        # In case of powers less than 1, number of terms need to be computed
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        if exp.is_positive:
+            if nu.is_integer:
+                #Reference: https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/ (only for integer order)
+                newn = ceiling(n/exp)
+                bn = besseli(nu, z)
+                a = ((-1)**(nu - 1)*log(z/2)*bn)._eval_nseries(x, n, logx, cdir)
+
+                b, c = [], []
+                o = Order(x**n, x)
+                r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
+                if r is S.Zero:
```

Comment:
The overall impelemtation looks correct to me. Can you let me know of a series expansion example where this change comes in though ?

Reply:
On master:
```
In [3]: besselk(2,x**2).series(x, n=2)
Out[3]: 
 ⎛ 2⎞
O⎝x ⎠
```

With this fix:
```
In [1]: besselk(2,x**2).series(x, n=2)
Out[1]: 
2     ⎛ 2⎞
── + O⎝x ⎠
 4        
x   
```

First sympy gets `x**2 -> O(x**2)`, but on putting this into the expansion of the Bessel function, we get `besselk(2,x**2) -> O(x**(-4))`. `Expr.series` notices that we did not get the requested order, and so increases the order of the intermediate series evaluations to get the answer shown above.

This also fixes the case `(x**2*besselk(2, x)).series(x ,n=2) == 2 + O(x**2)` (which I have added as a test).
Should I add the simpler case above to the tests as well?

Reply:
Correct, I do get your point now . We are missing out on some terms . I would say add the minimalistic example first and then you could add a couple more tests to top that. 
I realize we would find also similar problems in `bessely` too as I have pointed here (https://github.com/sympy/sympy/pull/25664#issuecomment-1713248527) . Would be interested in fixing that too ? Once we get this PR merged, that should be easy for you .

Reply:
Added the test.